docker/Dockerfile

```diff
@@ -41,13 +41,17 @@ RUN python3 -m venv /home/venv
 
 ENV PATH="/home/venv/bin:$PATH"
 
-RUN update-alternatives --install /usr/bin/python python /usr/bin/python3 1
-RUN update-alternatives --install /usr/local/bin/pip pip /usr/local/bin/pip3 1
+RUN update-alternatives --install /usr/bin/python python /usr/bin/python3 1 \
+    && update-alternatives --install /usr/local/bin/pip pip /usr/local/bin/pip3 1
 
-# This is only useful for cuda env 
+# This is only useful for cuda env
 RUN export USE_CUDA=1
 
-RUN pip install --no-cache-dir torch torchvision torchtext torchserve torch-model-archiver
+RUN if [ "$BASE_IMAGE" != "${BASE_IMAGE#*cuda:10.1}" ]; then \
+    pip install --no-cache-dir torch==1.6.0+cu101 torchvision==0.7.0+cu101 -f https://download.pytorch.org/whl/torch_stable.html; \
+    else \
+    pip install --no-cache-dir torch torchvision torchtext torchserve torch-model-archiver; \
+    fi
```

Comment:
```suggestion
ARG CUDA_VERSION

RUN TORCH_VER=$(curl --silent --location https://pypi.org/pypi/torch/json | python -c "import sys, json, pkg_resources; releases = json.load(sys.stdin)['releases']; print(sorted(releases, key=pkg_resources.parse_version)[-1])") && \
    TORCH_VISION_VER=$(curl --silent --location https://pypi.org/pypi/torchvision/json | python -c "import sys, json, pkg_resources; releases = json.load(sys.stdin)['releases']; print(sorted(releases, key=pkg_resources.parse_version)[-1])") && \
    if echo "$BASE_IMAGE" | grep -q "cuda:"; then \
        # Install CUDA version specific binary when CUDA version is specified as a build arg
        if [ "$CUDA_VERSION" ]; then \
            pip install --no-cache-dir torch==$TORCH_VER+$CUDA_VERSION torchvision==$TORCH_VISION_VER+$CUDA_VERSION -f https://download.pytorch.org/whl/torch_stable.html; \
        # Install the binary with the latest CUDA version support
        else \
            pip install --no-cache-dir torch torchvision; \
        fi \
    # Install the CPU binary
    else \
        pip install --no-cache-dir torch==$TORCH_VER+cpu torchvision==$TORCH_VISION_VER+cpu -f https://download.pytorch.org/whl/torch_stable.html; \
    fi
RUN pip install --no-cache-dir torchtext torchserve torch-model-archiver
```

Reply:
Added suggestion that adds:
* Checks for the CUDA_VERSION build argument and if present, installs the newest version of torch/torchvision with the supplied cuda version in the format of "cuda101", "cuda92".
* Pulls the CPU torch/torchvision binary when the BASE_IMAGE isn't explicitly passed in with the string "cuda". Before, the script was pulling the GPU-enabled binaries by default for CPU image builds. The resulting image is about 1GB smaller compared to the result of the previous build. (1.92GB -> 945MB)
* Supports legacy CUDA versions (e.g. cu92, etc.)

For the latest CUDA version that torch supports, users will have to specify a base image that supports it and NOT pass in the CUDA_VERSION argument. Made suggestion in the documentation too.

Reply:
Since the commands "pip install torch" will always be updated to pull the latest GPU-enabled binary, there wasn't a way to fully automate this without a way to query the latest CUDA version that PT supports. 
I initially parsed the BASE_IMAGE string to deduce the CUDA version and tried using the -f flag to search on download.pytorch.org for all GPU based builds. This didn't work because the latest PT1.6 + CUDA 102 is named as "torch-1.6.0" instead of "torch-1.6.0+cu102" and pip failed to find the correct binary.

Reply:
Attaching build logs:
[cpu_log.txt](https://github.com/pytorch/serve/files/5369574/cpu_log.txt)
[cuda101_log.txt](https://github.com/pytorch/serve/files/5369575/cuda101_log.txt)
[cuda102_log.txt](https://github.com/pytorch/serve/files/5369576/cuda102_log.txt)

Reply:
Thanks, @jeremiahschung. Merged all the recommendations :-)